bitcoin/wallet/gozer/tbcgozer/tbcgozer.go

```diff
@@ -0,0 +1,341 @@
+		Address:       addr.String(),
+		Start:         start,
+		Count:         count,
+	}
+
+	res, err := t.callTBC(ctx, defaultRequestTimeout, bur)
+	if err != nil {
+		return nil, err
+	}
+
+	buResp, ok := res.(*tbcapi.UTXOsByAddressResponse)
+	if !ok {
+		return nil, fmt.Errorf("not a utxos by address respose %T", res)
+	}
+
+	if buResp.Error != nil {
+		return nil, buResp.Error
+	}
+
+	return buResp.UTXOs, nil
+}
+
+func (t *tbcGozer) BlocksByL2AbrevHashes(ctx context.Context, hashes []chainhash.Hash) *gozer.BlocksByL2AbrevHashesResponse {
+	ksr := &tbcapi.BlocksByL2AbrevHashesRequest{
+		L2KeystoneAbrevHashes: hashes,
+	}
+
+	res, err := t.callTBC(ctx, defaultRequestTimeout, ksr)
+	if err != nil {
+		r := &gozer.BlocksByL2AbrevHashesResponse{
+			Error: protocol.Errorf("%v", err),
+		}
+		return r
+	}
+
+	bksr, ok := res.(*tbcapi.BlocksByL2AbrevHashesResponse)
+	if !ok {
+		r := &gozer.BlocksByL2AbrevHashesResponse{
+			Error: protocol.Errorf("not a keystone response %T", res),
+		}
+		return r
+	}
+	if bksr.Error != nil {
+		r := &gozer.BlocksByL2AbrevHashesResponse{
+			Error: bksr.Error,
+		}
+		return r
+	}
+	return gozer.TBC2Gozer(bksr)
+}
+
+func (t *tbcGozer) callTBC(pctx context.Context, timeout time.Duration, msg any) (any, error) {
+	log.Tracef("callTBC %T", msg)
+	defer log.Tracef("callTBC exit %T", msg)
+
+	if !t.Connected() {
+		return nil, errors.New("not connected to tbc")
+	}
+
+	bc := tbcCmd{
+		msg: msg,
+		ch:  make(chan any),
+	}
+
+	ctx, cancel := context.WithTimeout(pctx, timeout)
+	defer cancel()
+
+	// attempt to send
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case t.cmdCh <- bc:
+	default:
+		return nil, errors.New("tbc command queue full")
+	}
+
+	// Wait for response
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case payload := <-bc.ch:
+		if err, ok := payload.(error); ok {
+			return nil, err
+		}
+		return payload, nil
+	}
+
+	// Won't get here
+}
+
+func (t *tbcGozer) handleTBCWebsocketCall(ctx context.Context, conn *protocol.Conn) {
+	defer t.wg.Done()
+
+	log.Tracef("handleTBCWebsocketCall")
+	defer log.Tracef("handleTBCWebsocketCall exit")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case bc := <-t.cmdCh:
+			_, _, payload, err := tbcapi.Call(ctx, conn, bc.msg)
+			if err != nil {
+				log.Errorf("handleTBCWebsocketCall %T: %v", bc.msg, err)
+				select {
+				case bc.ch <- err:
+				default:
+				}
+			}
+			select {
+			case bc.ch <- payload:
+				log.Tracef("handleTBCWebsocketCall returned: %v", spew.Sdump(payload))
+			default:
+			}
+		}
+	}
+}
+
+func (t *tbcGozer) handleTBCWebsocketRead(ctx context.Context, conn *protocol.Conn) {
+	defer t.wg.Done()
+
+	log.Tracef("handleTBCWebsocketRead")
+	defer log.Tracef("handleTBCWebsocketRead exit")
+	pcc := conn.ConnectCount()
+	for {
+		_, _, _, err := tbcapi.ReadConn(ctx, conn)
+		if err != nil {
+			// See if we were terminated
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(5 * time.Second):
```

Comment:
This should really be an exponential backoff to avoid the issues we currently have with BFG.